### ISSUE TITLE
UndefinedBehaviorSanitizer: print-juniper.c:1407

### ISSUE THREAD

**Issue by gaintcome** [the-tcpdump-group/tcpdump#783]
Hi all,
I observed this UB in my fuzz testing. ASan does not complain about this issue, however, Intel MPX does.

print-juniper.c:1407:21: runtime error: index 8 out of bounds for type 'uint8_t [8]'
PoC:  [poc.zip](https://github.com/the-tcpdump-group/tcpdump/files/3408966/poc.zip)

Run (UBSan instrumented) : ``` ./tcpdump -r poc```
Output:
```
reading from file poc, link-type JUNIPER_MONITOR (Juniper Passive Monitor PIC), snapshot length 1600
print-juniper.c:1407:21: runtime error: index 8 out of bounds for type 'uint8_t [8]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior print-juniper.c:1407:21 in 
09:34:16.720640 Unknown Juniper DLT_ type 164: 
tcpdump: pcap_loop: bogus savefile header
```

Run (MPX instrumented) : ``` ./tcpdump -r poc```
Output:
```
reading from file poc, link-type JUNIPER_MONITOR (Juniper Passive Monitor PIC), snapshot length 1600
Saw a #BR! status 1 at 0x48e95a
Saw a #BR! status 1 at 0x48e95a
09:34:16.720640 Unknown Juniper DLT_ type 164: 
tcpdump: pcap_loop: bogus savefile header

```

**Comment by fxlb** [comment]
Hi,
Thanks for the report.

> ASan does not complain about this issue, however, Intel MPX does.

Are
`print-juniper.c:1407:21: runtime error: index 8 out of bounds for type 'uint8_t [8]'`
`SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior print-juniper.c:1407`
and
`Saw a #BR! status 1 at 0x48e95a`
`Saw a #BR! status 1 at 0x48e95a`
Linked ?

**Comment by fxlb** [comment]
Could you try this patch? It fixes undefined-behavior with UBSan.
```patch
diff --git a/print-juniper.c b/print-juniper.c
index 7ee82dc2..416b2f19 100644
--- a/print-juniper.c
+++ b/print-juniper.c
@@ -1399,6 +1399,11 @@ juniper_parse_header(netdissect_options *ndo,
                        lp->s,
                        l2info->cookie_len);
 
+            if (l2info->cookie_len > 8) {
+                nd_print_invalid(ndo);
+                return 0;
+            }
+
             if (l2info->cookie_len > 0) {
                 ND_TCHECK_LEN(p, l2info->cookie_len);
                 if (ndo->ndo_eflag)
```

**Comment by gaintcome** [comment]
Yes. That fixed the issue. Thanks!

**Comment by fxlb** [comment]
Fixed by 8c7221d39b44502999c78200d544665f13b4e331.

**Comment by fxlb** [comment]
backported to 4.9 with 45cb0ee8a0d6c247a41d8832c326740d17c99647.
